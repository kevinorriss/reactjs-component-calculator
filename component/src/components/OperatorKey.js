import React from 'react'
import PropTypes from 'prop-types'

class OperatorKey extends React.Component {
    constructor(props) {
        super(props)
        this.handleKeyDown = this.handleKeyDown.bind(this)
        this.state = {
            text: props.text,
            onClick: props.onClick
        }
    }

    componentDidMount() {
        window.addEventListener('keydown', this.handleKeyDown)
    }
    
    componentWillUnmount() {
        window.removeEventListener('keydown', this.handleKeyDown)
    }

    handleKeyDown(event) {
        if (event.key === this.state.text) {
            this.state.onClick(this.state.text)
        }
    }

    render() {
        return (
            <div className="button button--operator" onClick={() => this.state.onClick(this.state.text)}><div>{this.state.text}</div></div>
        )
    }
}

OperatorKey.propTypes = {
    onClick: PropTypes.func.isRequired,
    text: PropTypes.oneOf(['+', '-', '*', '/', '%']).isRequired
}

export default OperatorKey
