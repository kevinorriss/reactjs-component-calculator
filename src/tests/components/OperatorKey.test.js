import React from 'react'
import { shallow } from 'enzyme'
import renderer from 'react-test-renderer';
import OperatorKey from '../../../component/src/components/OperatorKey'

it('should render OperatorKey correctly', () => {
    const component = renderer.create(<OperatorKey text="+" onClick={jest.fn()}/>).toJSON()
    expect(component).toMatchSnapshot()
})

it('should call onClick when corresponding keyboard key pressed', () => {
    const onCick = jest.fn()
    const map = {}
    window.addEventListener = jest.fn((event, cb) => { map[event] = cb })
    shallow(<OperatorKey text="+" onClick={onCick}/>)
    map.keydown({ key: '+' })
    expect(onCick).toHaveBeenCalled()
})

it('should call onClick when clicked', () => {
    const onCick = jest.fn()
    const wrapper = shallow(<OperatorKey text="+" onClick={onCick}/>)
    wrapper.find('.button--operator').simulate('click')
    expect(onCick).toHaveBeenCalled()
})