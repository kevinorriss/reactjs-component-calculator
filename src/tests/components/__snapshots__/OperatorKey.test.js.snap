// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render OperatorKey correctly 1`] = `
<div
  className="button button--operator"
  onClick={[Function]}
>
  <div>
    +
  </div>
</div>
`;
